Add manual-trigger

paradigm opened the pull request:
This utility provides a quick means of submitting a jenkins job based on
a github pull request, bypassing sometimes unnecessary prow subsystems.

/cc stevekuznetsov

stevekuznetsov commented:
/cc @spxtr @kargakis 

The context here is we want to be able to run CI for our CI -- when a dev adds a new job, we use a tool like this to trigger the new jobs on a test cluster without having to involve all of the moving pieces around Prow, as that is not the system under test.

stevekuznetsov commented:
/ok-to-test

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+)
+
+type options struct {
+	githubapiurl string
```
style nit: camelcase to `githubApiUrl`

stevekuznetsov replied:
likewise with the rest of these

paradigm replied:
```
$ golint manual-trigger.go
manual-trigger.go:36:2: struct field githubApiUrl should be githubAPIURL
```

I don't feel strongly either way.  Should I do githubApiURL or githubAPIURL?
Perhaps I should work around it by naming this githubEndpoint

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+	fmt.Print("Querying GitHub for details... ")
+	pr, err := c.GetPullRequest(o.org, o.repo, o.num)
+	if err != nil {
+		fmt.Print("failed!\n")
```
these are not necessary

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+	prowjob := pjutil.NewProwJob(spec)
+	jc := jenkins.NewClient(o.jenkinsurl, &ac)
+
+	fmt.Print("Submitting to Jenkins... ")
```
`Println`

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+	err = jc.Build(&prowjob)
+	if err != nil {
+		fmt.Print("failed!\n")
+		log.Fatal("Error submitting to Jenins: ", err)
```
`s/Jenins/Jenkins`

stevekuznetsov replied:
style nit: we use:

```go
log.Fatalf("Error submitting to Jenins: %v\n", err)
```

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+	jc := jenkins.NewClient(o.jenkinsurl, &ac)
+
+	fmt.Print("Submitting to Jenkins... ")
+	err = jc.Build(&prowjob)
```
Looks like this method only uses the `Spec` and `Metadata.Name` -- let's refactor to take the minimal set of arguments

paradigm replied:
What do you have in mind with the refactor?  Change `Build()` to only take `Spec` and `Metadata.Name` and rework the rest of the codebase which calls it, perhaps as a separate PR?

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+		},
+	}
+
+	job := config.Presubmit{
```
Construct the `ProwJobSpec` literal here, we're not gaining much from the `PresubmitSpec()` call

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,190 @@
+		log.Fatal("cannot read file specified by --jbtoken: %v", err)
+	}
+
+	c := github.NewClient(ghtoken, o.githubapiurl)
```
Try to hoist as much as possible before this call -- the Jenkins client set-up, etc -- so we don't use tokens that would get wasted when something else down the line fails

paradigm replied:
The only things I see that could fail after this depend on it:

- gc.GetPullRequest can fail, but needs gc
- gc.GetRef can fail, but it needs gc
- jc.Build can fail, but it needs gc.GetRef which needs gc

I can move jenkins.NewClient() up before there, but I don't see the benefit.  Not sure if I'm missing something.

stevekuznetsov commented:
/lint

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		log.Fatalf("Unable to get reference to pull request %v", err)
+	}
+	fmt.Println("done")
```
These `Print` statements are what i meant before -- if it's critical information for run-time debugging, use a logger and log them. But if it's just a comment like `"done"` I don't think we should keep them in the final version of this tool

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		log.Fatalf("Error submitting to Jenkins %v", err)
+	}
+	fmt.Println("done")
```
For instance, you may want to print out the job name/number and the parameters that were passed here.

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+
+	pr, err := gc.GetPullRequest(o.org, o.repo, o.num)
+	if err != nil {
+		log.Fatalf("Unable to get information on pull request (org=%q repo=%q pr-num=%d): %v", o.org, o.repo, o.num, err)
```
We usually use `Printf("%s/%s#%d", org, repo, num)` to format this triplet of info

kargakis replied:
The use of Fatalf seems fine here.

stevekuznetsov replied:
Errr I just meant the way the three pieces of information were formatted

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+
+	err = jc.BuildFromSpec(&spec, o.jobName)
+	if err != nil {
+		log.Fatalf("Error submitting Jenkins job (org=%q repo=%q pr-num=%d ref=%q) with job name %q to %q: %v", o.org, o.repo, o.num, ref, o.jobName, o.jenkinsEndpoint, err)
```
Ideally here we'd list the actual parameters we passed in to the job -- maybe try calling `EnvForSpec()` and giving us either a machine-readable output or a table?

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	err = jc.BuildFromSpec(&spec, o.jobName)
+	if err != nil {
```
If `err` is not necessary, you can keep it out of the larger scope

```go
if err := foo(); err != nil {

}
```

stevekuznetsov commented on prow/jenkins/jenkins.go:
```diff
@@ -214,17 +214,23 @@ func (c *Client) doRequest(method, path string) (*http.Response, error) {
-// Build triggers a Jenkins build for the provided ProwJob. The name of
-// the ProwJob is going to be used as the buildId parameter that will help
-// us track the build before it's scheduled by Jenkins.
+// Thin wrapper around BuildFromSpec which allows it to be called with a full
```
Doc did not need to change

stevekuznetsov commented on prow/jenkins/jenkins.go:
```diff
@@ -214,17 +214,23 @@ func (c *Client) doRequest(method, path string) (*http.Response, error) {
+	return c.BuildFromSpec(&pj.Spec, pj.Metadata.Name)
+}
+
+// Build triggers a Jenkins build for the provided ProwJobSpec. The name of the
```
`s/Build/BuildFromSpec/`

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.githubEndpoint, "githubapiurl", "https://api.github.com", "GitHub's API endpoint.")
```
`github-endpoint` to be consistent with elsewhere.

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.githubEndpoint, "githubapiurl", "https://api.github.com", "GitHub's API endpoint.")
+	flag.StringVar(&o.jenkinsEndpoint, "jenkinsurl", "", "Jenkins url.")
```
`[jenkind-url](https://github.com/kubernetes/test-infra/blob/4dd7e4dcd350e767e1418b2b815a5914e7943daa/prow/cmd/jenkins-operator/main.go#L40)`

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+	flag.StringVar(&o.githubEndpoint, "githubapiurl", "https://api.github.com", "GitHub's API endpoint.")
+	flag.StringVar(&o.jenkinsEndpoint, "jenkinsurl", "", "Jenkins url.")
+
+	flag.StringVar(&o.githubToken, "ghtoken", "", "Path to file containing GitHub token.")
```
[`github-token-file`](https://github.com/kubernetes/test-infra/blob/4dd7e4dcd350e767e1418b2b815a5914e7943daa/prow/cmd/jenkins-operator/main.go#L46)

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+	flag.StringVar(&o.jenkinsEndpoint, "jenkinsurl", "", "Jenkins url.")
+
+	flag.StringVar(&o.githubToken, "ghtoken", "", "Path to file containing GitHub token.")
+	flag.StringVar(&o.jenkinsBearerToken, "jbtoken", "", "Path to file containing Jenkins Bearer token.")
```
[`jenkins-bearer-token-file`](https://github.com/kubernetes/test-infra/blob/4dd7e4dcd350e767e1418b2b815a5914e7943daa/prow/cmd/jenkins-operator/main.go#L43)

Do we want to support basic auth?

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+
+	_, err := url.Parse(o.githubEndpoint)
+	if o.githubEndpoint == "" || err != nil {
+		return fmt.Errorf("must specify a valid URL to --githubApiUrl")
```
Can you include the error in the message?

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+
+	_, err = url.Parse(o.jenkinsEndpoint)
+	if o.jenkinsEndpoint == "" || err != nil {
+		return fmt.Errorf("must specify a valid URL to --jenkinsurl")
```
Same question as above

kargakis commented on prow/jenkins/jenkins.go:
```diff
@@ -214,17 +214,23 @@ func (c *Client) doRequest(method, path string) (*http.Response, error) {
+// Build triggers a Jenkins build for the provided ProwJobSpec. The name of the
+// ProwJob is going to be used as the buildId parameter that will help us track
+// the build before it's scheduled by Jenkins.
+func (c *Client) BuildFromSpec(spec *kube.ProwJobSpec, name string) error {
```
`name` here is `buildId` for Jenkins.

kargakis replied:
`name` here is `buildId` for Jenkins so maybe rename to `buildId`.

kargakis commented:
We need to also make this work for kubernetes prowjobs.

cjwagner commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+	if o.jenkinsBearerToken == "" {
+		return fmt.Errorf("empty --jbtoken")
+	}
+	if o.jenkinsBearerToken == "" {
```
This should be `o.jobName`

cjwagner commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,171 @@
+	if o.num <= 0 {
+		return fmt.Errorf("empty or invalid --num")
+	}
+	if o.repo == "" {
```
s/repo/org/

fejta commented:
I would also like a way to manually trigger Prow jobs. Does this help here?

stevekuznetsov commented:
@fejta with or without that trigger being picked up by the Prow controllers? This effort was mostly to circumvent that.

paradigm commented:
Per kargakis' comments about consistency I found other code in the code base and reworked things accordingly.  I also added BasicAuth support - not sure why I failed to put that in originally.

I'm not sure the error output in case of an issue with the message to Jenkins is what stevekuznetsov had in mind; feedback there may be needed.

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+		log.Fatalf("Unable to get information on pull request %s/%s#%d: %v", o.org, o.repo, o.num, err)
+	}
+
+	ref, err := gc.GetRef(o.org, o.repo, "heads/"+pr.Base.Ref)
```
Don't you already have all required info in pr.Base and pr.Head?

kargakis commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+		MaxConcurrency: 1,
+	}
+
+	if err = jc.BuildFromSpec(&spec, o.jobName); err != nil {
```
We should also make it easy to run bare k8s pods. Can you make the same change you did on the jenkins client for [`ProwJobToPod`](https://github.com/kubernetes/test-infra/blob/2aff63d011fa1a636be14a4530e89047579f29b1/prow/pjutil/pjutil.go#L122)? We could read a pod spec from the prow config on Github.

kargakis replied:
I think this would also address @fejta's request for triggering prowjobs manually.

stevekuznetsov replied:
I agree this would be useful functionality but we should tackle it as a follow-up.

spxtr commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+
+	_, err := url.Parse(o.githubEndpoint)
+	if o.githubEndpoint == "" || err != nil {
+		return fmt.Errorf("bad --github-endpoint provided: %v", err)
```
if it's `""` then add a message saying it must be set.

spxtr commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+
+	_, err = url.Parse(o.jenkinsURL)
+	if o.jenkinsURL == "" || err != nil {
+		return fmt.Errorf("bad --jenkins-url provided: %v", err)
```
likewise

spxtr commented:
Can we build this into `cmd/mkpj`? Seems like there is some overlap.

spxtr commented:
Wait, nevermind. This doesn't actually make a ProwJob, it just goes straight to Jenkins.

BenTheElder commented:
/area jenkins

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+		log.Fatal(err)
+	}
+
+	ac := jenkins.AuthConfig{}
```
```
// TODO(kargakis): dry this out
```

stevekuznetsov commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	if err = jc.BuildFromSpec(&spec, o.jobName); err != nil {
+		log.Println("Submitting the following to Jenkins:")
```
We also want the job name/number in a click-able link that will take me to the job running (if that's possible, unless it just dumps it into the queue and we can't return that, in which case we should at least give the link to the server+job)

paradigm commented:
I believe this resolves all mentioned issues baring requests to extend the scope.  Provided there are no objections, I'd like to see if I can get this in just to trigger Jenkins.  We can add support for triggering prowjobs and such things later once this is pulled in.

If I missed anything, let me know.

stevekuznetsov commented:
@cjwagner @spxtr any more comments?

cjwagner commented on experiment/manual-trigger/manual-trigger.go:
```diff
@@ -0,0 +1,199 @@
+		MaxConcurrency: 1,
+	}
+
+	if false {
```
Please delete this conditional since it is never entered.

paradigm replied:
Woops >.<

paradigm commented:
@cjwagner updated accordingly, let me know if there's anything else

stevekuznetsov commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) [CI] Improve the workflow around pipeline jobs for CI devs](https://trello.com/c/2pzxVODg/362-3-ci-improve-the-workflow-around-pipeline-jobs-for-ci-devs)

cjwagner commented:
/lgtm

kargakis commented:
@paradigm please open an issue so we can track extending this tool
